### ISSUE TITLE
Feature request: Post and Pre commands

### ISSUE THREAD

**Issue by jocado** [unixsurfer/anycast_healthchecker#18]
It would be really awesome, if you can run some additional arbitrary shell command-line when a route is enabled or disabled.

The idea is to have some options like this per check definition:

`pre_disabled_cmd`
`post_disabled_cmd`
`pre_enabled_cmd`
`post_enabled_cmd`

I think this would be a very useful addition.

My use case currently is, I would like to be able to actually down the interface as well as withdraw the route for some check and hosts, and of course up the interface then before the relevant route is advertised again.

I would be happy to look at it at some point and submit a merge request, but I'm short on time for it right now.  If it magically appeared in the next release, it would be most awesome! :)

Thanks for a very useful utility.

**Comment by unixsurfer** [comment]
I understand the use case. What do you want to happen if any of those cmds fail?

**Comment by jocado** [comment]
Thanks for you're quick response.

I think if the `pre` commands fail they should abort the change of state, and the `post` commands should just log the error.

I suppose you could make the behaviours configurable [ if you set strict mode aborts state change, else it carries on ], but that would make the feature more complex to implement.

What do you think?

**Comment by unixsurfer** [comment]
> Thanks for you're quick response.
> 
> I think if the `pre` commands fail they should abort the change of state, and the `post` commands should just log the error.
> 

We determine the **status** of the service based on the return code of the script set in *check_cmd*. We want to keep this, very fundamental principle, unchanged. Moreover, if we have two different checks (think them as two different brains) to decide the status of the service then we are going to, unnecessarily, complicate our life.  Moreover, the **status** is determined before we run those `pre` cmds and it doesn't sound good  to change the status **again** in case on of them fails. Thus, I don't want the failure of `pre` cmds to change the **status** of the service. I prefer to log an error message.

I would like to mention two things about those `pre` cmds and I would like to hear your opinion.

We need to set timeouts on `pre` cmds as we need to finish all the work before the next check. We currently use *check_timeout* to set the timeout for  *check_cmd*. Shall we use the same  for `pre` cmds, or shall we introduce  *pre_check_timeout*?

Moreover, I think we should be able to call those `pre` cmds many times without getting different outcome(similar to idempotent HTTP methods), otherwise we may run into weird situations. What do you think?

**Comment by jocado** [comment]
I understand your concerns with over-complicating status changes, and tracking too much status.

Before we get into your last questions, I've had an idea that may simply implementation, while still satisfying my particular use case at least.

Is it possible set  `bird_reconfigure_cmd` on a per service check basis?

If it was, then I could simply change it to a script or command string that ran the commands I wanted for a particular check.

If we extend that idea further, you could have an `advertise_cmd` and a `withdraw_cmd` per check [ up/down commands], which both default to `bird_reconfigure_cmd` as set in defaults, but are then configurable per check.

This seems simpler to me.  You only have one exec or shell call to worry about still, and no extra timeouts.

Does that sound better, or do you prefer the `pre` and `post` idea?

**Comment by unixsurfer** [comment]
> I understand your concerns with over-complicating status changes, and tracking too much status.
> 
> Before we get into your last questions, I've had an idea that may simply implementation, while still satisfying my particular use case at least.
> 
> Is it possible set `bird_reconfigure_cmd` on a per service check basis?
> 
> If it was, then I could simply change it to a script or command string that ran the commands I wanted for a particular check.

But, how then will we run different commands (pre and post)? 
BTW: we have a [specific logic](https://github.com/unixsurfer/anycast_healthchecker/blob/master/anycast_healthchecker/utils.py#L701) to determine if `bird reconfigure` succeeded, so we can't use the same setting.

> 
> If we extend that idea further, you could have an `advertise_cmd` and a `withdraw_cmd` per check [ up/down commands], which both default to `bird_reconfigure_cmd` as set in defaults, but are then configurable per check.

I like this approach better. I propose to have those per service check, specific timeouts( `advertise_cmd_timeout` and a `withdraw_cmd_timeout` ) and have them unset by default. We check the exit code and simply report an error if they fail, and capture to log the STDERR if they squeak anything to it. But, without any modifications for `bird_reconfigure_cmd`, so we will also run it if the status changes.

What do you think?

**Comment by jocado** [comment]
> But, how then will we run different commands (pre and post)?

For that I was thinking, I was think you could just use a script, which runs commands and logic in any sequence you desire.

> BTW: we have a specific logic to determine if bird reconfigure succeeded, so we can't use the same setting.

OK - that's fair :)

> I like this approach better. I propose to have those per service check, specific timeouts( advertise_cmd_timeout and a withdraw_cmd_timeout ) and have them unset by default. We check the exit code and simply report an error if they fail, and capture to log the STDERR if they squeak anything to it. But, without any modifications for bird_reconfigure_cmd, so we will also run it if the status changes.

That sounds good, but it brings us back to ordering again.

For my use case, up/down and interface, I _should_ up the interface before the advertise and down the interface after the withdraw.

Although, thinking about it, I suppose it is technically feasible to do that in any order.  The bird config change is for a white list, so it's then up to bird to add or remove routes as interlaces appear or disappear.  So perhaps order isn't a real problem for me, I was just trying to keep things "tidy".

If you make or want order to configurable, what's nicer for you to implement?  Run the `advertsie_cmd` and `withdraw_cmd` before or after the `bird_reconfigure_cmd` ?

**Comment by unixsurfer** [comment]
> > But, how then will we run different commands (pre and post)?
> 
> For that I was thinking, I was think you could just use a script, which runs commands and logic in any sequence you desire.

Maybe we can add `custom_bird_reconfigure_cmd` and when it is set we only run this and not the `bird_reconfigure_cmd`. Which implies that the custom script will run bird reconfigure and return exit code 1 if it fails(so it has to duplicate the custom logic we have). We can also pass an argument to inform the script that we run it during the removal or addition of the route
```
custom_script down
custome_script up
```
> 
> > BTW: we have a specific logic to determine if bird reconfigure succeeded, so we can't use the same setting.
> 
> OK - that's fair :)
> 
> > I like this approach better. I propose to have those per service check, specific timeouts( advertise_cmd_timeout and a withdraw_cmd_timeout ) and have them unset by default. We check the exit code and simply report an error if they fail, and capture to log the STDERR if they squeak anything to it. But, without any modifications for bird_reconfigure_cmd, so we will also run it if the status changes.
> 
> That sounds good, but it brings us back to ordering again.
> 
> For my use case, up/down and interface, I _should_ up the interface before the advertise and down the interface after the withdraw.
> 
> Although, thinking about it, I suppose it is technically feasible to do that in any order. The bird config change is for a white list, so it's then up to bird to add or remove routes as interlaces appear or disappear. So perhaps order isn't a real problem for me, I was just trying to keep things "tidy".
> 

what happens if we do bird reconfigure if the network interface is disabled? Have you tried that? Maybe bird will complain.

> If you make or want order to configurable, what's nicer for you to implement? Run the `advertsie_cmd` and `withdraw_cmd` before or after the `bird_reconfigure_cmd` ?

**Comment by jocado** [comment]
> Maybe we can add custom_bird_reconfigure_cmd and when it is set we only run this and not the bird_reconfigure_cmd. Which implies that the custom script will run bird reconfigure and return exit code 1 if it fails(so it has to duplicate the custom logic we have). We can also pass an argument to inform the script that we run it during the removal or addition of the route
> ```
> custom_script down
> custom_script up
> ```
> 

That sounds good to me  :+1: 

..as long as we can specify `custom_bird_reconfigure_cmd` per check [ which I think you are suggesting ].

If someone uses `custom_bird_reconfigure_cmd` it's up to them to make sure it robust and exits with an appropriate exit status.

> what happens if we do bird reconfigure when the network interface is disabled? Have you tried that? Maybe bird will complain.

This will probably depend how your bird config is written.  Our config simply exports interface routes which bird finds in the kernel routing table [ with some extra filtering rules applied].  So, in our case it's fine, and has been tested many times :)

**Comment by unixsurfer** [comment]
> > Maybe we can add custom_bird_reconfigure_cmd and when it is set we only run this and not the bird_reconfigure_cmd. Which implies that the custom script will run bird reconfigure and return exit code 1 if it fails(so it has to duplicate the custom logic we have). We can also pass an argument to inform the script that we run it during the removal or addition of the route
> > ```
> > custom_script down
> > custom_script up
> > ```
> 
> That sounds good to me +1
> 
> ..as long as we can specify `custom_bird_reconfigure_cmd` per check [ which I think you are suggesting ].
> 

correct.
> If someone uses `custom_bird_reconfigure_cmd` it's up to them to make sure it robust and exits with an appropriate exit status.
> 
> > what happens if we do bird reconfigure when the network interface is disabled? Have you tried that? Maybe bird will complain.
> 
> This will probably depend how your bird config is written. Our config simply exports interface routes which bird finds in the kernel routing table [ with some extra filtering rules applied]. So, in our case it's fine, and has been tested many times smile

OK then we need to mention this in the doc.

So, to sum up.
- Introduce a new option with the name `custom_bird_reconfigure_cmd` and invoke it if it is set by the user
- `custom_bird_reconfigure_cmd` should be set per service check
- `bird_reconfigure_cmd` is skipped if `custom_bird_reconfigure_cmd` is set
- we log error if exit code is not 0 and log STDERR, if there is any
- failure of `custom_bird_reconfigure_cmd` does not change the state
- add `timeout_custom_bird_reconfigure_cmd`

Did I forget anything ?

**Comment by jocado** [comment]
That sounds great.

Thanks very much for your interest and support :smile: 

I'm very willing to help with any testing.  Please let me know if I can do anything.

**Comment by unixsurfer** [comment]
I would like to first emit a new release, we haven't made a release for several months and there are several commits, and then work on this.
Now, I have to be very honest with you. I have absolutely zero free time and as a result, I can not promise I will work on it the coming days/weeks. Having said that, I will try to see if I can develop a beta version of this in 30mins during a commute.

**Comment by jocado** [comment]
Please don't worry about timescale.  Your project, and help, is very much appreciated.

Just let me know when you want me to test something.

**Comment by unixsurfer** [comment]
I pushed 3588f1c5aec626ba0091a623052cdd6d09b6c9e9 into custom_bird_reconfigure_cmd branch.
Please give it a try by adding `custom_bird_reconfigure_cmd` in the config file of  a service check. The default timeout is 2 seconds and you can overwrite it with `custom_bird_reconfigure_cmd_timeout`, oops just realized I should have used `timeout_custom_bird_reconfigure_cmd`

**Comment by unixsurfer** [comment]
@jocado  I have pushed some updates to the branch, please make sure you use the latest

**Comment by jocado** [comment]
Awesome - I should be able to test this tomorrow :+1:

**Comment by zerthimon** [comment]
Awesome feature request and I'm definitely going to use it when it gets implemented. I'm using RIP to announce my prefix and BFD for fast convergence. But when a prefix is withdrawn it just stops being announced and it takes additional time for the infrastructure to converge. So i'd use a post-down hook (`birdc disable bfd1`) to stop the BFD and pre-up hook (`birdc enable bfd`) to start it for the immediate convergence.

**Comment by unixsurfer** [comment]
@zerthimon you can already test it by using custom_bird_reconfigure_cmd branch

**Comment by zerthimon** [comment]
@unixsurfer Thanks, I'm going test it and let you know the result.

**Comment by zerthimon** [comment]
@unixsurfer it didn't work
I specified `custom_bird_reconfigure_cmd = /usr/sbin/birdc disable bfd1` but in log `/usr/sbin/birdc disable bfd1 down` and also bird was not reloaded.

Also where do I specify `/usr/sbin/birdc enable bfd1` when healthcheck passes ?

**Comment by unixsurfer** [comment]
@zerthimon, from the documentation of **custom_bird_reconfigure_cmd**

> A custom command to trigger a reconfiguration of Bird daemon. This overwrites the value of bird_reconfigure_cmd and bird6_reconfigure_cmd settings, which allows to use a custom command to trigger a reconfiguration of Bird daemon after and IP prefix is either added to or removed from Bird configuration. It allows to use a different command per service as it is a setting per service check. If return code is not a zero value then an error is logged together with STDERR of the command, if there is any. anycast-healthchecker passes one argument to the command, which is up when IP prefix is added or down when is removed.

It does not work because we pass  *down* to **birdc** command line tool, which does not obviously know that argument. You aren't supposed to set **custom_bird_reconfigure_cmd** to birdc tool.
The whole idea with **custom_bird_reconfigure_cmd* is to run a script that will do the reconfiguration and possible run other cmds. Since we pass either *up* or *down, the script can do the appropriate thing when a service is marked up or down.
I guess the documentation wasn't clear enough, sorry for this. Do you have any ideas on how we can improve that specific part of the documentation?

**Comment by zerthimon** [comment]
@unixsurfer oops, I've not read that part of the documentation after I checked out branch, my bad.
Let me try again. Be back with results.

**Comment by jocado** [comment]
Apologies @unixsurfer , I've been pulled onto something urgent, so I haven't been able to test yet.  I will do as soon as I can.

**Comment by zerthimon** [comment]
@unixsurfer tested the custom_bird_reconfigure_cmd and it works.
I used the following script:
```
#!/bin/bash

case ${!#} in
     down)      
	  /usr/sbin/birdc configure
          /usr/sbin/birdc disable bfd1
          ;;
     up)      
	  /usr/sbin/birdc configure
          /usr/sbin/birdc enable bfd1
          ;;
     *)
          echo "Unknown status: ${!#}"
          exit 1
          ;;
esac
```
The convergence is much faster now, thanks!

**Comment by unixsurfer** [comment]
> @unixsurfer tested the custom_bird_reconfigure_cmd and it works.
> I used the following script:
> 
> ```
> #!/bin/bash
> 
> case ${!#} in
>      down)      
> 	  /usr/sbin/birdc configure
>           /usr/sbin/birdc disable bfd1
>           ;;
>      up)      
> 	  /usr/sbin/birdc configure
>           /usr/sbin/birdc enable bfd1
>           ;;
>      *)
>           echo "Unknown status: ${!#}"
>           exit 1
>           ;;
> esac
> ```
> 
> The convergence is much faster now, thanks!



Thanks a lot for the feedback.

**Comment by unixsurfer** [comment]
> There is a problem I found during my testing:
> If the script is not executable the program just crashes to the command line:
> 
> ```
> 2019-01-21 17:03:34,756 anycast-healthchecker[19143] INFO     MainThread reconfiguring BIRD by running custom command /etc/anycast-healthchecker/scripts/bird_reconfigure.sh down
> Traceback (most recent call last):
>   File "/usr/local/bin/anycast-healthchecker", line 10, in <module>
>     sys.exit(main())
>   File "/usr/local/lib/python3.5/dist-packages/anycast_healthchecker/main.py", line 98, in main
>     checker.run()
>   File "/usr/local/lib/python3.5/dist-packages/anycast_healthchecker/healthchecker.py", line 228, in run
>     run_custom_bird_reconfigure(operation)
>   File "/usr/local/lib/python3.5/dist-packages/anycast_healthchecker/utils.py", line 1224, in run_custom_bird_reconfigure
>     stderr=subprocess.PIPE)
>   File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
>     restore_signals, start_new_session)
>   File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
>     raise child_exception_type(errno_num, err_msg)
> PermissionError: [Errno 13] Permission denied```
> ```

This is because I handle only *FileNotFoundError* and *subprocess.TimeoutExpired* exception, I swap  *FileNotFoundError* with *OSError* in order to cover both cases. Pushing  a commit into the branch.

Thanks for the report and I am sorry for this.

**Comment by unixsurfer** [comment]
Pushed https://github.com/unixsurfer/anycast_healthchecker/commit/d298ba4ee2a82748f735bb3ff39fea43c82f3f0e

**Comment by unixsurfer** [comment]
@jocado and @zerthimon: Have you managed to find any time to test this? I would like to emit a new release next week and it would be great if we could include this as well.

**Comment by jocado** [comment]
@unixsurfer I have just got back some time to devote to this now.  I will try and get it done today and report back :)

**Comment by zerthimon** [comment]
@unixsurfer yep, I've tested the branch up to this commit https://github.com/unixsurfer/anycast_healthchecker/commit/d298ba4ee2a82748f735bb3ff39fea43c82f3f0e and it works for me, see above.

**Comment by jocado** [comment]
Sorry for the delay, I've only just got the packaging sorted out [ it's been a while since we upgraded ].   I have it all sorted nicely now, and I will move onto actually testing tomorrow!

**Comment by jocado** [comment]
@unixsurfer Thanks for your patience, I have now been able to test this feature.

It seems to work well, and as designed :+1: 

However, unfortunately it doesn't quite satisfy my use case.  The problem is because I want to down the interface with the VIP, when it comes to starting re-advertising the prefix again, anycast-healthchecker refuses because:
```
status DOWN because 1.2.3.4/32 isn't assigned to loopback interface
```

Would it be possible to make the check for the interlace configurable ?

In our set-up, having the interface up or down doesn't effect birds runtime health, it just carries on and either adds the prefix to be advertised or not [ depending on whether the interface is up or down ].

This will illustrate what I need to achieve:
```
#!/bin/bash
set -e
case $1 in
	up)
		/sbin/ifup vip:test
		/usr/sbin/birdc configure
	;;
	down)
		/usr/sbin/birdc configure
		/sbin/ifdown vip:test
	;;
	*)
		echo "Only up/down supported, got: ${1}"
	;;
esac
```

**Comment by unixsurfer** [comment]
Thanks for the feedback. Please set **ip_check_disabled** to true to avoid the check against the IP Address assignment

**Comment by jocado** [comment]
@unixsurfer Apologies - I didn't realise that option existed already! :smile: 

Perfect, I can now use it as intended for us.  It works great - thanks very much for your support :+1:

**Comment by unixsurfer** [comment]
@jocado, in your bash script you may want to exit with **1**  code in case a wrong input is given in order to get an error in the anycast-healthchecker.log and be aware of it.

Now, let's see if I can find the time to emit a new release(0.9.0) this or the coming week.

**Comment by jocado** [comment]
@unixsurfer Yes you're correct, and I will be :+1:   this was just a quick script for testing.

I look forward to the new release :smiley:

**Comment by unixsurfer** [comment]
0.9.0 version is out!
